x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,2399 @@
+
+Both of these methods use `internalRuleToAPIResponse` internally to transform rules into our API response schema, so the normalization will come out of the box when executing that transformation.
+
+Also, in order to allow the endpoint to export both custom **and** prebuilt rules, we need to update the logic and remove the checks that we currently do server-side in both of these methods, and which filter out prebuilt rules from the response payload:
+
+- **Bulk Actions** - `POST /rules/_bulk_action` with the **export action**
+- **Export Rules** - `POST /rules/_export`
+
+_Source: [x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/export/get_export_by_object_ids.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/export/get_export_by_object_ids.ts)_
+
+```ts
+// [... file continues above ...]
+
+export const getRulesFromObjects = async (
+  rulesClient: RulesClient,
+  savedObjectsClient: RuleExecutorServices['savedObjectsClient'],
+  objects: Array<{ rule_id: string }>,
+  logger: Logger
+): Promise<RulesErrors> => {
+  const chunkedObjects = chunk(objects, 1024);
+  const filter = [...]
+  const rules = await findRules({[...]});
+
+  const alertsAndErrors = objects.map(({ rule_id: ruleId }) => {
+    const matchingRule = rules.data.find((rule) => rule.params.ruleId === ruleId);
+    if (
+      matchingRule != null
+      && hasValidRuleType(matchingRule)
+      // && matchingRule.params.immutable !== true  <--- Check which should be removed
+    ) {
+      return {
+        statusCode: 200,
+        rule: transformRuleToExportableFormat(internalRuleToAPIResponse(matchingRule)),
+      };
+    } else {
+      return {
+        statusCode: 404,
+        missingRuleId: { rule_id: ruleId },
+      };
+    }
+  });
+
+// [... file continues below ...]
+```
+
+Notice that the **Bulk Actions** - `POST /rules/_bulk_action` does not perform a **dry run** in order to filter out prebuilt rules when the user runs an **export** action; this is the only check that is performed.
+
+*Source: [x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/export/get_export_all.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/export/get_export_all.ts)*
+```ts
+// [... file continues above ...]
+export const getExportAll = async (
+  // [...]
+): Promise<{
+  rulesNdjson: string;
+  // [...]
+}> => {
+  // const ruleAlertTypes = await getNonPackagedRules({ rulesClient }); <---- current code
+  const allRules = await getRules({ rulesClient }); // <------ get all rules now
+  const rules = transformAlertsToRules(allRules);
+
+// [... file continues below ...]
+```
+
+### Importing rules
```

Comment:
Some updates per today's discussion at the Simplified Protections WG:

@jpdjere made a great point that we need to allow users to specify the version on import in order for upgrades to work correctly. If a user exports a prebuilt rule from one cluster at version 3 and imports it to another cluster where an updated package has been installed with version 4 of that rule, and we import the rule without specifying the version, we won't be able to easily tell if the imported rule is version 4 customized or version 3 un-customized. If we import the rule and treat it as version 4 customized, but it should have been version 3 un-customized, then upgrading the rule to version 5 would require the user to merge "their customizations" that they didn't actually make. If we instead allow the rule to be imported as version 3 un-customized, the upgrade flow is easier as there are no customizations to worry about merge and we can simply fast forward to version 5 when they want to upgrade.

With that case in mind, I think the combination of `rule_id` + `version` would be the unique identifier we use on rule import to load and populate the additional `prebuilt` fields like updated at, `isCustomized`, etc based on the security-rule assets. With the addition of `version`, we can pick the right asset from the history instead of always using the latest one as I had imagined previously.

The core requirement that I had in mind when I proposed that the import API should not allow users to set the `prebuilt` fields was that the `prebuilt` fields must always be in sync with the associated `security-rule` asset - we need to do validation at the API boundaries so our application within those boundaries can always assume that the data in those fields is reliable. But maybe it would be convenient to allow users to include `prebuilt` fields in the ndjson they import to mark which rules should be linked to security-rule assets, and during the import process we could load the prebuilt rules package (if it is not already loaded) and either overwrite or reject rules in the import JSON with `prebuilt` fields that don't match up with what's in the package.

Reply:
@marshallmain 

> But maybe it would be convenient to allow users to include prebuilt fields in the ndjson they import to mark which rules should be linked to security-rule assets

Not sure I understand why we'd need the data from the `prebuilt` field if we have the `rule_id` and the `version` and are fetching the rule asset based on those two values. If the `prebuilt` field can be completely recalculated from there, why would we allow `prebuilt` to be sent in the payload?

Reply:
If `prebuilt` can be passed in the payload then we don't need an extra parameter on the API, separate from the rule import data, to control whether or not we try to link imported rules to prebuilt rule assets - each rule's JSON tells us if it should be linked or not. None of the data like `isCustomized` or `sourceUpdatedAt` from the `prebuilt` object would be copied directly into the created rule, but the existence of the `prebuilt` object would signal the user's intent to import a rule as a prebuilt rule instead of a custom rule that just coincidentally shares a rule_id and version with a prebuilt rule.

With the `ruleSource` changes we talked about, if `ruleSource.type: 'external'` (or `ruleSource` is missing and `immutable: true`) then we'd fetch the rest of the `ruleSource` data from the security rule asset. If `ruleSource.type: 'internal` (or `ruleSource` is missing and `immutable: false`) then we wouldn't need to fetch any other data from the security rule assets. I think with `ruleSource.type` now the idea is a bit clearer, whereas before we would look for `prebuilt` but recalculate all the fields in it so the `prebuilt` data seems redundant (we would just be looking for the existence of the object), with `ruleSource` we keep `ruleSource.type` and recalculate the rest of the values if type is `external`.